src/sentry/services/hybrid_cloud/organization_mapping/__init__.py

```diff
@@ -0,0 +1,71 @@
+from abc import abstractmethod
+from dataclasses import dataclass
+from datetime import datetime
+from typing import Any, Optional
+
+from django.utils import timezone
+
+from sentry.models.user import User
+from sentry.services.hybrid_cloud import InterfaceWithLifecycle, silo_mode_delegation, stubbed
+from sentry.silo import SiloMode
+
+
+@dataclass(frozen=True, eq=True)
+class APIOrganizationMapping:
+    id: int = -1
+    organization_id: int = -1
+    slug: str = ""
+    region_name: str = ""
+    date_created: datetime = timezone.now()
+    verified: bool = False
+    customer_id: Optional[str] = None
+
+
+class OrganizationMappingService(InterfaceWithLifecycle):
+    @abstractmethod
+    def create(
+        self,
+        *,
+        user: User,
+        organization_id: int,
+        slug: str,
+        region_name: str,
+        idempotency_key: Optional[str] = "",
+        customer_id: Optional[str],
+    ) -> APIOrganizationMapping:
+        """
+        This method returns a new or recreated OrganizationMapping object.
+        Will raise IntegrityError if the slug already exists.
```

Comment:
When you say that it raises an IntegrityError however, that seems odd to me given the idempotency key.

When I invoke create twice with the same idempotency key, does it throw an integrity error?  Or only when two different idempotency keys are used?

Reply:
idempotency_key will be used to re-allow creation of an (organization_id, slug) pair

Reply:
In fact, can we remove the `IntegrityError` expectation here?  I think it's a little dangerous -- we won't necessarily, in fact, be sending up an IntegrityError when we swap out for RPC.  It's possible it will be a general error type.

Reply:
In fact, come to think of it, I ponder aloud if it makes sense for the signature here to be `Optional[` instead of using errors to signal that the create did not succeed.  Honestly, it's not exception or a bug at all if concurrent creates happen, that is expected part of the application's behavior.  Callers should be simply be mindful and check for the presence of a created organization at call site?

Reply:
That said, these are nits and not critical blockers.